src/query/block/meta.go

```diff
@@ -139,25 +140,51 @@ type ResultMetadata struct {
 	// this result.
 	FetchedMetadataCount int
 	// MetricNames is the set of unique metric tag name values across all series in this result.
-	MetadataByName map[string]*ResultMetricMetadata
+	// External users must access via `ByName(name)`.
+	metadataByName map[string]*ResultMetricMetadata
+}
+
+// AddNamespace adds a namespace to the namespace set, initializing the underlying map if necessary.
+func (m *ResultMetadata) AddNamespace(namespace string) {
+	if m.namespaces == nil {
+		m.namespaces = make(map[string]struct{})
+	}
+	m.namespaces[namespace] = struct{}{}
+}
+
+// GetNamespaces returns an array representing the set of namespaces added via AddNamespace.
+func (m ResultMetadata) GetNamespaces() []string {
+	if m.namespaces == nil {
+		return []string{}
+	}
+	namespaces := []string{}
```

Comment:
nit: could alloc to the len of namespaces

Reply:
I'll follow up with an optimization PR that contains this and the internal tag lookup search short-circuit.